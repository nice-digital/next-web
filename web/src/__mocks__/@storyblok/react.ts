export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
