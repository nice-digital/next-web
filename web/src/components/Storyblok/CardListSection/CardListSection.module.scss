@use '~/styles/colours';
@use '@nice-digital/nds-core/scss/layout';
@use '@nice-digital/nds-core/scss/media-queries';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/typography';
@use '@nice-digital/nds-core/scss/utils';
@use '~/styles/spacing' as *;

.cardSection {
  @include layout.fullWidth();
  background-color: colours.$card-section-background;
}

.cardSectionLargeSpacing {
  padding: $section-large-spacing 0;
}

.cardSectionMediumSpacing {
  padding: $section-medium-spacing 0;
}

.cardSectionSmallSpacing {
  padding: $section-small-spacing 0;
}

.heading {
  margin-top: 0;
}

.leadText {
  p {
    @include typography.font-size(lead);
  }
}
.cardSectionListIntro{
  margin-bottom: utils.rem(spacing.$large);
}
.transparent {
  background: transparent;
}

.container {
  @include layout.fullWidthContainer();
}
