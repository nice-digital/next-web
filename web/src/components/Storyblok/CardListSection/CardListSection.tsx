import classnames from "classnames";
import Link from "next/link";

import { Card, type CardHeadingLinkProps } from "@nice-digital/nds-card";

import { StoryblokRichText } from "@/components/Storyblok/StoryblokRichText/StoryblokRichText";
import { CardListSectionStoryblok } from "@/types/storyblok";
import { fieldHasValidContent, resolveStoryblokLink } from "@/utils/storyblok";
import { toTitleCase } from "@/utils/string";

import styles from "./CardListSection.module.scss";

export interface CardListSectionProps {
	blok: CardListSectionStoryblok;
}

export const CardListSection: React.FC<CardListSectionProps> = ({
	blok,
}: CardListSectionProps) => {
	const {
		heading,
		headingLevel,
		leadText,
		secondaryLeadText,
		verticalPadding = "medium",
		theme,
		cards,
	} = blok;

	const HeadingElement = `h${headingLevel}` as keyof JSX.IntrinsicElements;
	const transparentClass = theme === "subtle" ? undefined : styles.transparent;
	const verticalPaddingClass = `cardSection${toTitleCase(
		verticalPadding
	)}Spacing`;
	const richTextHasContent = leadText ? fieldHasValidContent(leadText) : false;
	const secondaryLeadTextProcessed =
	secondaryLeadText && fieldHasValidContent(secondaryLeadText) ? secondaryLeadText : null;

	return (
		<section
			className={classnames(
				styles.cardSection,
				styles[verticalPaddingClass],
				transparentClass
			)}
		>
			<div className={styles.container}>
				{heading || richTextHasContent ? (
					<div className={styles.cardSectionListIntro}>
						{heading && (
							<HeadingElement className={styles.heading}>
								{heading}
							</HeadingElement>
						)}
						{richTextHasContent && (
							<StoryblokRichText
								content={leadText}
								className={styles.leadText}
							/>
						)}
						{richTextHasContent && secondaryLeadTextProcessed && (
							<StoryblokRichText
								content={secondaryLeadTextProcessed}
								className={styles.leadTextSecondary}
							/>
						)}
					</div>
				) : undefined}

				<ul className="list list--unstyled">
					{cards.map(({ heading, body, link, _uid }) => {
						let cardLink: CardHeadingLinkProps | undefined = undefined;
						const resolvedLink = link ? resolveStoryblokLink(link) : undefined;
						if (resolvedLink?.url) {
							cardLink = {
								destination: resolvedLink.url,
								elementType: resolvedLink.isInternal ? Link : "a",
								method: "href",
							};
						}

						return (
							<Card
								elementType="li"
								headingText={heading}
								link={cardLink || undefined}
								key={_uid}
							>
								{body}
							</Card>
						);
					})}
				</ul>
			</div>
		</section>
	);
};
