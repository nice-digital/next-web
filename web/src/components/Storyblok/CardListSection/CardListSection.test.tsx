import { render, screen } from "@testing-library/react";

import { CardListSection, type CardListSectionProps } from "./CardListSection";

const mockCardListSectionProps: CardListSectionProps = {
	blok: {
		heading: "Mock promo box title",
		leadText: {
			type: "Mock rich text",
		},
		headingLevel: "2",
		verticalPadding: "medium",
		theme: "subtle",
		cards: [
			{
				heading: "card heading",
				body: " card summary",
				link: {
                    id: "",
                    url: "https://nice.org.uk/guidance/ta10",
                    linktype: "url",
                    fieldtype: "multilink",
                    cached_url: "https://nice.org.uk/guidance/ta10",
                },
				component: "cardListSectionItem",
				_uid: "123456877"
			},
		],
		component: "cardListSection",
        _uid: "1234578",
	},
};
describe("Promo box component", () => {
	it("should match snapshot", () => {
		const { container } = render(
			<CardListSection {...mockCardListSectionProps} />
		);
		expect(container).toMatchSnapshot();
	});
});
