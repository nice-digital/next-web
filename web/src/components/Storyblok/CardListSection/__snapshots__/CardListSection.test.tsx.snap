// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardListSection component should match snapshot when card section list intro is not present 1`] = `
<div>
  <section
    class="cardSection cardSectionMediumSpacing"
  >
    <div
      class="container"
    >
      <ul
        class="list list--unstyled"
      >
        <li
          class="card"
          data-component="card"
        >
          <header
            class="card__header"
          >
            <p
              class="card__heading"
            >
              <a
                href="https://nice.org.uk/guidance/ta10"
              >
                Mock card heading 1
              </a>
            </p>
          </header>
          <p
            class="card__summary"
          >
            Mock card summary 1
          </p>
        </li>
      </ul>
    </div>
  </section>
</div>
`;

exports[`cardListSection component should match snapshot with default values 1`] = `
<div>
  <section
    class="cardSection cardSectionMediumSpacing"
  >
    <div
      class="container"
    >
      <div
        class="cardSectionListIntro"
      >
        <h2
          class="heading"
        >
          Mock cardListSection title
        </h2>
        <div
          class="storyblokRichTextWrapper leadText"
          data-testid="storyblok-rich-text"
        >
          <p>
            Mock cardListSection lead text
          </p>
        </div>
        <div
          class="storyblokRichTextWrapper leadTextSecondary"
          data-testid="storyblok-rich-text"
        >
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut felis leo, fermentum quis posuere non, viverra at nulla.
          </p>
        </div>
      </div>
      <ul
        class="list list--unstyled"
      >
        <li
          class="card"
          data-component="card"
        >
          <header
            class="card__header"
          >
            <p
              class="card__heading"
            >
              <a
                href="https://nice.org.uk/guidance/ta10"
              >
                Mock card heading 1
              </a>
            </p>
          </header>
          <p
            class="card__summary"
          >
            Mock card summary 1
          </p>
        </li>
      </ul>
    </div>
  </section>
</div>
`;

exports[`cardListSection component should match snapshot with transparent theme and large padding 1`] = `
<div>
  <section
    class="cardSection cardSectionLargeSpacing transparent"
  >
    <div
      class="container"
    >
      <div
        class="cardSectionListIntro"
      >
        <h2
          class="heading"
        >
          Mock cardListSection title
        </h2>
        <div
          class="storyblokRichTextWrapper leadText"
          data-testid="storyblok-rich-text"
        >
          <p>
            Mock cardListSection lead text
          </p>
        </div>
        <div
          class="storyblokRichTextWrapper leadTextSecondary"
          data-testid="storyblok-rich-text"
        >
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut felis leo, fermentum quis posuere non, viverra at nulla.
          </p>
        </div>
      </div>
      <ul
        class="list list--unstyled"
      >
        <li
          class="card"
          data-component="card"
        >
          <header
            class="card__header"
          >
            <p
              class="card__heading"
            >
              <a
                href="https://nice.org.uk/guidance/ta10"
              >
                Mock card heading 1
              </a>
            </p>
          </header>
          <p
            class="card__summary"
          >
            Mock card summary 1
          </p>
        </li>
      </ul>
    </div>
  </section>
</div>
`;
